Fixing End-to-End Tests with action.js and waitFor.js for protractor_utils/users.js fixes part of #10798

ashutoshc8101 opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes part of #10798
2. This PR does the following:  Fix end-to-end tests with action.js and waitFor.js for protractor_utils/users.js

## Essential Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The linter/Karma presubmit checks have passed locally on your machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [x] The PR is made from a branch that's **not** called "develop".

## Proof that changes are correct

I tested the changes to users.js file using following tests:
![Screenshot from 2021-02-12 20-00-35](https://user-images.githubusercontent.com/24855641/107787746-84b52f80-6d75-11eb-98c1-9f29239f163d.png)
![Screenshot from 2021-02-12 20-03-00](https://user-images.githubusercontent.com/24855641/107788131-0efd9380-6d76-11eb-95c6-2816459d7b8a.png)

<!--
Add videos/screenshots of the user-facing interface to demonstrate that the changes made in this PR work correctly.
-->

## PR Pointers

- Make sure to follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Contributing-code-to-Oppia#instructions-for-making-a-code-change).
- Oppiabot will notify you when you don't add a PR_CHANGELOG label. If you are unable to do so, please @-mention a code owner (who will be in the Reviewers list), or ask on [Gitter](https://gitter.im/oppia/oppia-chat).
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays.
- Never force push. If you do, your PR will be closed.
- Oppiabot can assign anyone for review/help if you leave a comment like the following: "{{Question/comment}} @{{reviewer_username}} PTAL"
- Some of the e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-your-build-fails).

ashutoshc8101 committed:
Fixing End-to-End Tests with action.js and waitFor.js for protractor_utils/users.js partly fixes (#10798)

Linting issue fixed

linting fix

ashutoshc8101 commented:
@U8NWXD @kevintab95 Please review.

kevintab95 commented on core/tests/protractor_utils/users.js:
```diff
@@ -19,15 +19,17 @@
-
+var action = require('./action.js');
 var AdminPage = require('./AdminPage.js');
+const { browser } = require('protractor');
```
This is not required, I think.

ashutoshc8101 replied:
I tested without the import, it is working. I will remove the import.

kevintab95 commented on core/tests/protractor_utils/users.js:
```diff
@@ -96,17 +102,15 @@ var _completeSignup = async function(username, manualNavigation = true) {
 var completeLoginFlowFromStoryViewerPage = async function(email, username) {
-  await login(email, false, false);
-  await _completeSignup(username, false);
+  await login(email, true);
+  await _completeSignup(username);
```
Are these expected changes?

ashutoshc8101 replied:
If I don't set manualNavigation to true, Following happens:
![Screenshot from 2021-02-13 11-57-08](https://user-images.githubusercontent.com/24855641/107843627-d64fcf80-6df2-11eb-865a-8e4ff59a9c35.png)
![Screenshot from 2021-02-13 11-57-58](https://user-images.githubusercontent.com/24855641/107843628-d8b22980-6df2-11eb-9d11-184e3195cba6.png)
for this test:
![Screenshot from 2021-02-13 11-59-06](https://user-images.githubusercontent.com/24855641/107843633-e667af00-6df2-11eb-8386-1f83ec0a812b.png)

isSuperAdmin should be false, that was a mistake, It will change it.

kevintab95 commented on core/tests/protractor_utils/users.js:
```diff
@@ -72,8 +76,10 @@ var login = async function(
 
 var logout = async function() {
   var driver = browser.driver;
+  browser.waitForAngularEnabled(false);
```
Why do we need this?

ashutoshc8101 replied:
If i dont use this, test fails with following error:
![Screenshot from 2021-02-13 12-02-30](https://user-images.githubusercontent.com/24855641/107843735-76a5f400-6df3-11eb-93df-282fa3b82ebb.png)
![Screenshot from 2021-02-13 12-02-44](https://user-images.githubusercontent.com/24855641/107843740-79084e00-6df3-11eb-8c57-32c825ae4113.png)

kevintab95 commented on core/tests/protractor_utils/users.js:
```diff
@@ -19,15 +19,17 @@
   // Use of element is not possible because the login page is non-angular.
   // The full url is also necessary.
   var driver = browser.driver;
+  await browser.waitForAngularEnabled(false);
```
Why do we need this?

ashutoshc8101 replied:
If I don't add this, the following test fails:
![Screenshot from 2021-02-13 12-06-50](https://user-images.githubusercontent.com/24855641/107843807-16fc1880-6df4-11eb-91d9-aeb587d661b5.png)
![Screenshot from 2021-02-13 12-07-02](https://user-images.githubusercontent.com/24855641/107843809-195e7280-6df4-11eb-9fa9-62b8561acca5.png)

kevintab95 replied:
Hi @ashutoshc8101, just a general comment, please don't resolve comments by yourself. It makes it harder for reviewers to see previous comments.

kevintab95 replied:
Hm, it fails because using action.* expects the page to be an Angular page but the login screen is not Angular currently. Using `waitForAngularEnabled` will disable the waitForAngular check, but it also makes tests flaky so we don't suggest using it.


Here's what I would suggest. Remove `waitForAngularEnabled` here and below that you added and remove usage of "action" in the login / logout screen. I think it should be possible to use action in lines 104-106 without `waitForAngularEnabled` but please confirm this by running any e2e test suite that uses the corresponding method on your local machine.

If it is not possible to use action there, I think we can exclude "users.js" from the list of files that need to use action. Maybe you can add a detailed comment at the top of this file about why we can't use "action".

kevintab95 commented on core/tests/protractor_utils/users.js:
```diff
@@ -46,16 +48,18 @@ var login = async function(
-    await (await driver.findElement(protractor.By.name('admin'))).click();
-    let adminCheckboxStatus = await driver.findElement(
-      protractor.By.name('admin')).getAttribute('checked');
+    var testAdminButton = element(protractor.By.name('admin'));
```
```suggestion
    var adminCheckbox = element(protractor.By.name('admin'));
```

This is a checkbox and not a button.

ashutoshc8101 replied:
Fixed.

kevintab95 commented on core/tests/protractor_utils/users.js:
```diff
@@ -72,8 +76,10 @@ var login = async function(
+  browser.waitForAngularEnabled(false);
   await driver.get(general.SERVER_URL_PREFIX + general.LOGIN_URL_SUFFIX);
-  await (await driver.findElement(protractor.By.id('submit-logout'))).click();
+  var testSubmitButton = element(protractor.By.id('submit-logout'));
```
Maybe we can call it logoutButton?

ashutoshc8101 replied:
Fixed.

kevintab95 reviewed (changes requested):
Thanks @ashutoshc8101! Left a few comments. PTAL!

U8NWXD reviewed (approved):
I'm approving since kevintab95 has already reviewed the e2e tests

ashutoshc8101 committed:
Requested changes are implemented

ashutoshc8101 commented:
@kevintab95 Requested changes have been implemented.

sajalasati commented:
Hey @ashutoshc8101! If the PR is ready for review kindly leave a PTAL comment for the reviewer. Oppiabot will then assign the reviewer. (Always ensure the correct person is assigned to the PR before waiting for a review. You can read more about the instructions [here](https://github.com/oppia/oppia/wiki/Contributing-code-to-Oppia#address-review-comments-until-all-reviewers-give-lgtm-looks-good-to-me)) Thanks!

ashutoshc8101 commented:
PTAL

ashutoshc8101 commented:
@kevintab95 PTAL

ashutoshc8101 commented:
@kevintab95 Please review.

ashutoshc8101 committed:
Merge remote-tracking branch 'upstream/develop' into FixingETEUsers

kevintab95 reviewed (changes requested):
Thanks @ashutoshc8101! Left a comment, PTAL!

ashutoshc8101 committed:
Merge remote-tracking branch 'upstream/develop' into FixingETEUsers

ashutoshc8101 committed:
action use is reverted and appropiate comments added

ashutoshc8101 commented:
@kevintab95 PTAL.
I reverted the use of action.* in login and logout function. 
It was possible to use action without waitForAngularEnabled in lines 104-106.
I have added appropriate comments regarding the use of action in login and logout functions.

I ran the following tests to check:
![Screenshot from 2021-02-19 14-05-28](https://user-images.githubusercontent.com/24855641/108480284-c20d4600-72bc-11eb-86e4-817a9cf442d9.png)

All ran successfully:
![Screenshot from 2021-02-19 14-05-11](https://user-images.githubusercontent.com/24855641/108480299-c9ccea80-72bc-11eb-86af-635a55880813.png)

In the future, I will not resolve comments by myself.

ashutoshc8101 committed:
Merge remote-tracking branch 'upstream/develop' into FixingETEUsers

kevintab95 commented on core/tests/protractor_utils/users.js:
```diff
@@ -87,26 +88,22 @@ var _completeSignup = async function(username, manualNavigation = true) {
     // as a client side navigation and the tests fail since Angular is
     // not found due to the navigation interfering with protractor's
     // bootstrapping.
-    await browser.waitForAngularEnabled(false);
     await browser.get('/signup?return_url=http%3A%2F%2Flocalhost%3A9001%2F');
-    await browser.waitForAngularEnabled(true);
```
Let's keep this.

ashutoshc8101 replied:
Done.

kevintab95 commented on core/tests/protractor_utils/users.js:
```diff
@@ -87,26 +88,22 @@ var _completeSignup = async function(username, manualNavigation = true) {
 };
 
 var completeLoginFlowFromStoryViewerPage = async function(email, username) {
-  await login(email, false, false);
-  await _completeSignup(username, false);
```
Please add these back.

ashutoshc8101 replied:
Done.

kevintab95 reviewed (changes requested):
Thanks @ashutoshc8101! Left few comments. PTAL!

ashutoshc8101 committed:
Request changes

ashutoshc8101 commented:
@kevintab95 Thank you for your review.
I have implemented the requested changes.
PTAL.

kevintab95 reviewed (approved):
LGTM, thanks @ashutoshc8101!

U8NWXD commented:
@ashutoshc8101 the learner flow test failed 3 times in a row with this failure:

```
1) Learner dashboard functionality should visit the collection player and play the correct collection
  - Failed: Register User button is not clickable.
  Wait timed out after 10051ms
```

This e2e test has been pretty reliable recently, so it's unusual that it failed 3 times in a row. Are you able to run this test locally? You can do so like this:

```
python -m scripts.run_e2e_tests --suite learner --prod_env
```

ashutoshc8101 committed:
Merge remote-tracking branch 'upstream/develop' into FixingETEUsers

ashutoshc8101 committed:
Merge remote-tracking branch 'upstream/develop' into FixingETEUsers

ashutoshc8101 committed:
E2E Test fixed

ashutoshc8101 commented:
@U8NWXD PTAL.
The test was failing because length of username exceeds 30 characters.
So I shortened the username.
![Screenshot from 2021-02-23 20-30-06](https://user-images.githubusercontent.com/24855641/108864250-0596e580-7618-11eb-966d-c90c83349848.png)
Now tests are passing:
![Screenshot from 2021-02-23 20-44-04](https://user-images.githubusercontent.com/24855641/108864252-0760a900-7618-11eb-8b0c-4bc97d1645b7.png)
